File: Luogu/P3864 [USACO1.2]命名那个数字 Name That Number.cpp
#include<bits/stdc++.h>

using namespace std;

int main() {
    string num;
    cin >> num;
    const string nums = "23456789";
    const char names[8][3] = {{'A', 'B', 'C'}, {'D', 'E', 'F'}, {'G', 'H', 'I'}, {'J', 'K', 'L'},
                        {'M', 'N', 'O'}, {'P', 'R', 'S'}, {'T', 'U', 'V'}, {'W', 'X', 'Y'}};
    vector<string> acceptable_names = {};
    for (int i = 0; i < 4167; i++) {
        string temp;
        cin >> temp;
        acceptable_names.push_back(temp);
    }

    vector<string> new_acceptable_names = {};
    for (string & acceptable_name : acceptable_names){
        if (acceptable_name.length() == num.length()){
            new_acceptable_names.push_back(acceptable_name);
        }
    }
    acceptable_names = new_acceptable_names;
    new_acceptable_names = {};

    for (int i = 0; i < static_cast<int>(num.length()); i++){
        for (string & acceptable_name : acceptable_names){
            bool flag = false;
            for (int j = 0; j < 3; j++){
                if (acceptable_name[i] == names[nums.find(num[i])][j]) {
                    flag = true;
                    break;
                }
            }
            if (flag){
                new_acceptable_names.push_back(acceptable_name);
            }
        }
        acceptable_names = new_acceptable_names;
        new_acceptable_names = {};
    }

    if (acceptable_names.empty()){
        cout << "NONE" << endl;
    }else{
        for (string & name: acceptable_names){
            cout << name << endl;
        }
    }

    return 0;
}

File: Luogu/P3864 [USACO1.2]命名那个数字 Name That Number.py
num = input()
num_name = {
    '2': ['A', 'B', 'C'], '3': ['D', 'E', 'F'], '4': ['G', 'H', 'I'],
    '5': ['J', 'K', 'L'], '6': ['M', 'N', 'O'], '7': ['P', 'R', 'S'],
    '8': ['T', 'U', 'V'], '9': ['W', 'X', 'Y']
}
acceptable_names = []
for _ in range(4617):
    acceptable_names.append(input())

new_acceptable_names = []
for i in range(len(acceptable_names)):
    if len(acceptable_names[i]) == len(num):
        new_acceptable_names.append(acceptable_names[i])
acceptable_names = new_acceptable_names.copy()
new_acceptable_names = []

for i in range(len(num)):
    for j in range(len(acceptable_names)):
        if acceptable_names[j][i] in num_name[num[i]]:
            new_acceptable_names.append(acceptable_names[j])
    acceptable_names = new_acceptable_names.copy()
    new_acceptable_names = []

if len(acceptable_names) == 0:
    print('NONE')
else:
    for name in acceptable_names:
        print(name)
